frame/merkle-mountain-range/src/primitives.rs

```diff
@@ -0,0 +1,392 @@
+	fn using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F, _compact: bool) -> R {
+		codec::Encode::using_encoded(self, f)
+	}
+}
+
+/// An element representing either raw leaf data or a hash.
+///
+/// See [Compact] to see how it may be used in practice to reduce the size
+/// of proofs in case multiple [LeafDataProvider]s are composed together.
+///
+/// [DataOrHash::hash] method is calculating the hash of this element in it's compact form,
+/// so should be used instead of hashing the encoded form (which will always be non-compact).
+#[derive(RuntimeDebug, Clone, PartialEq)]
+pub enum DataOrHash<H: traits::Hash, L> {
+	/// A terminal leaf node, storing arbitrary content.
+	Data(L),
+	/// An inner node - always just a hash.
+	Hash(H::Output),
+}
+
+impl<H: traits::Hash, L> From<L> for DataOrHash<H, L> {
+	fn from(l: L) -> Self {
+		Self::Data(l)
+	}
+}
+
+mod encoding {
+	use super::*;
+
+	/// A helper type to implement [codec::Codec] for [DataOrHash].
+	#[derive(codec::Encode, codec::Decode)]
+	enum Either<A, B> {
+		Left(A),
+		Right(B),
+	}
+
+	impl<H: traits::Hash, L: FullLeaf> codec::Encode for DataOrHash<H, L> {
+		fn encode_to<T: codec::Output>(&self, dest: &mut T) {
+			match self {
+				Self::Data(l) => l.using_encoded(
+					|data| Either::<&[u8], &H::Output>::Left(data).encode_to(dest), false
+				),
+				Self::Hash(h) => Either::<&[u8], &H::Output>::Right(h).encode_to(dest),
+			}
+		}
+	}
+
+	impl<H: traits::Hash, L: FullLeaf> codec::Decode for DataOrHash<H, L> {
+		fn decode<I: codec::Input>(value: &mut I) -> Result<Self, codec::Error> {
+			let decoded: Either<Vec<u8>, H::Output> = Either::decode(value)?;
+			Ok(match decoded {
+				Either::Left(l) => DataOrHash::Data(L::decode(&mut &*l)?),
+				Either::Right(r) => DataOrHash::Hash(r),
+			})
+		}
+	}
+}
+
+impl<H: traits::Hash, L: FullLeaf> DataOrHash<H, L> {
+	/// Retrieve a hash of this item.
+	///
+	/// Depending on the node type it's going to either be a contained value for [DataOrHash::Hash]
+	/// node, or a hash of SCALE-encoded [DataOrHash::Data] data.
+	pub fn hash(&self) -> H::Output {
+		match *self {
+			Self::Data(ref leaf) => leaf.using_encoded(<H as traits::Hash>::hash, true),
+			Self::Hash(ref hash) => hash.clone(),
+		}
+	}
+}
+
+/// A composition of multiple leaf elements with compact form representation.
+///
+/// When composing together multiple [LeafDataProvider]s you will end up with
+/// a tuple of `LeafData` that each element provides.
+///
+/// However this will cause the leaves to have significant size, while for some
+/// use cases it will be enough to prove only one element of the tuple.
+/// That's the rationale for [Compact] struct. We wrap each element of the tuple
+/// into [DataOrHash] and each tuple element is hashed first before constructing
+/// the final hash of the entire tuple. This allows you to replace tuple elements
+/// you don't care about with their hashes.
```

Comment:
I am not sure I understand the advantages of `Compact` over using a inner type to the leaf that is a tuple of hashes, and in a second step map the tuple elements to their respective DataOrHash.

Reply:
Well, the thing is that we want to keep `LeafDataProvider`s clean. Imagine we have these 3 providers:
```
type Providers = (
  frame_system::Module<T> /// returns a parent block hash
  parachain_bridge::Module<T> // returns a list (or a tree) of parachain headers
  bridge_messages::Module<T> // returns a list of messages to send over the bridge
);
```
We implement the `LeafDataProvider` for a tuple as well, so it returns something like:
```
type Leaf = (H256, Vec<ParaHead>, Vec<BridgeMessage>);
```
To make it possible to efficiently prove say `.0` without the need of providing the entire leaf data (meaning full lists of `.1` and `.2`), we rather want something like this as the leaf data:
```
type CompactLeaf = (
  DataOrHash<H256>,
  DataOrHash<Vec<ParaHead>>,
  DataOrHash<Vec<BridgeMessage>>
)
```
To obtain that we would either need to change each of the original `Providers`s to return `DataOrHash` or wrap each one of them into something that wraps the output.
Instead we allow one to do:
```
type CompactProviders = Compact<Providers>;
```
and `CompactProviders` type automatically implements `LeafDataProvider` which returns the `CompactLeaf` struct that we expect.